File: contracts/generator/game-parts.ts
/**
 * Golden vectors for the part rules: which files of a game folder (on Drive,
 * on the SSD or on a camera card) are the game, and in what order.
 */
import { DEFAULT_TOLERANCE, vectorCase, type VectorFile } from "./vector";

import { MAX_SOURCES } from "@/features/games/validation";
import { selectGameParts } from "@/features/ingest/parts";

function partsCase(name: string, fileNames: string[]) {
  return vectorCase(name, "selectGameParts", { fileNames }, (i) => {
    const result = selectGameParts(i.fileNames);
    // The reason is English log text; a port reports the same outcome in
    // its own words, so only the outcome is pinned.
    return result.kind === "invalid" ? { kind: "invalid" } : result;
  });
}

/** One more part than a game may have: 101 one-chapter recordings. */
function tooManyRecordings(): string[] {
  return Array.from(
    { length: MAX_SOURCES + 1 },
    (_, i) => `GX01${String(i + 1).padStart(4, "0")}.MP4`,
  );
}

export function buildGameParts(): VectorFile {
  return {
    contract: "game-parts",
    description:
      "The game parts among a folder's file names, in play order. Three schemes " +
      "count, matched case-insensitively: halbzeit<N> and viertel<N> (.mp4 or " +
      ".mov), and GoPro chapters G[XH]<CC><NNNN>.MP4 ordered by recording NNNN, " +
      "then chapter CC. Other files are ignored and listed in code-unit order. A " +
      "folder mixing schemes, repeating or skipping a part, or holding more than " +
      "maxParts parts is invalid (its English reason is left out).",
    reference: ["src/features/ingest/parts.ts"],
    tolerance: DEFAULT_TOLERANCE,
    constants: { maxParts: MAX_SOURCES },
    cases: [
      partsCase("GoPro chapters in any listing order", [
        "GX020042.MP4",
        "GX030042.MP4",
        "GX010042.MP4",
      ]),
      partsCase("two recordings play in recording order", [
        "GX010043.MP4",
        "GX020042.MP4",
        "GX010042.MP4",
      ]),
      partsCase("H.264 chapters", ["GH020007.MP4", "GH010007.MP4"]),
      partsCase("HEVC and H.264 recordings are one scheme", [
        "GX010008.MP4",
        "GH010007.MP4",
      ]),
      partsCase("lower-case names", ["gx020042.mp4", "gx010042.mp4"]),
      partsCase("a camera card's side files are ignored", [
        "GX010042.MP4",
        "GX010042.THM",
        "GL010042.LRV",
        "GX020042.MP4",
        "GL020042.LRV",
      ]),
      partsCase("other files in a game folder are ignored", [
        "notes.txt",
        "GX010042.MP4",
        "Tor.MOV",
        "cover.jpg",
      ]),
      partsCase("exported halves", ["Halbzeit 2.mp4", "halbzeit1.mp4"]),
      partsCase("exported quarters", [
        "Viertel_3.MOV",
        "viertel1.mp4",
        "Viertel-2.mp4",
        "viertel4.mov",
      ]),
      partsCase("a folder without parts", ["notes.txt", "cover.jpg"]),
      partsCase("an empty folder", []),
      partsCase("mixed schemes are invalid", ["halbzeit1.mp4", "GX010042.MP4"]),
      partsCase("a repeated part is invalid", [
        "halbzeit1.mp4",
        "Halbzeit 1.mp4",
      ]),
      partsCase("a missing chapter is invalid", [
        "GX010042.MP4",
        "GX030042.MP4",
      ]),
      partsCase("a recording without its first chapter is invalid", [
        "GX020042.MP4",
      ]),
      partsCase("a missing half is invalid", ["halbzeit2.mp4"]),
      partsCase(
        "more parts than a game may have is invalid",
        tooManyRecordings(),
      ),
    ],
  };
}

File: mac/HockeyKit/Sources/HockeyCore/GameParts.swift
/// Which files of a game folder are the game, and in which order.
///
/// Port of `src/features/ingest/parts.ts`, pinned by
/// `contracts/vectors/game-parts.json`. Three naming schemes count, matched
/// case-insensitively:
///
/// - `halbzeit<N>`: an exported half (`halbzeit1.mp4`, `Halbzeit 2.mp4`),
/// - `viertel<N>`: an exported quarter (`Viertel1.mp4`),
/// - GoPro chapters `GX<CC><NNNN>.MP4` (HEVC) or `GH<CC><NNNN>.MP4` (H.264):
///   chapter `CC` of recording `NNNN`, ordered by recording, then chapter.
///
/// Everything else is ignored. A folder that mixes schemes, repeats a part,
/// skips a number or holds more parts than a game may have is invalid rather
/// than guessed at.

/// The most parts one game may have (`MAX_SOURCES` on the web).
public let maxGameParts = 100

/// The naming scheme a folder's parts follow.
public enum PartScheme: String, Sendable, Comparable {
    case halbzeit
    case viertel
    case gopro

    public static func < (lhs: PartScheme, rhs: PartScheme) -> Bool {
        lhs.rawValue < rhs.rawValue
    }
}

/// One part a problem is about: a half or quarter number, or a GoPro chapter of
/// a recording.
public struct PartLabel: Equatable, Sendable {
    public let scheme: PartScheme
    /// GoPro recording number; 0 for the exported schemes.
    public let recording: Int
    /// Half, quarter or GoPro chapter number.
    public let index: Int

    public init(scheme: PartScheme, recording: Int, index: Int) {
        self.scheme = scheme
        self.recording = recording
        self.index = index
    }
}

/// Why a folder's parts cannot be ordered into one game.
public enum GamePartsProblem: Equatable, Sendable {
    /// The folder mixes naming schemes (sorted by name).
    case mixedSchemes([PartScheme])
    /// The folder holds more than `maxGameParts` parts.
    case tooManyParts(count: Int)
    /// The same part appears under two file names.
    case repeated(PartLabel, first: String, second: String)
    /// A part the sequence needs is not there.
    case missing(PartLabel)
}

/// What `selectGameParts` found in a folder.
public enum GamePartsResult: Equatable, Sendable {
    /// The folder's parts in play order, plus the files that were ignored.
    case parts(scheme: PartScheme, parts: [String], ignored: [String])
    /// No file in the folder is a game part.
    case none(ignored: [String])
    /// Parts exist but cannot be ordered into one game.
    case invalid(GamePartsProblem)
}

/// One recognised part and the key it sorts by.
private struct Part {
    let name: String
    let label: PartLabel
}

private func parsePart(_ name: String) -> Part? {
    let exported = #/(?i)(halbzeit|viertel)[\s_\-]*([0-9]{1,2})\.(?:mp4|mov)/#
    if let match = try? exported.wholeMatch(in: name),
       let scheme = PartScheme(rawValue: match.1.lowercased()),
       let index = Int(match.2)
    {
        return Part(name: name, label: PartLabel(scheme: scheme, recording: 0, index: index))
    }
    let gopro = #/(?i)G[XH]([0-9]{2})([0-9]{4})\.mp4/#
    if let match = try? gopro.wholeMatch(in: name),
       let chapter = Int(match.1),
       let recording = Int(match.2)
    {
        return Part(
            name: name,
            label: PartLabel(scheme: .gopro, recording: recording, index: chapter)
        )
    }
    return nil
}

/// Orders strings by UTF-16 code units, the order JavaScript's default sort and
/// `<` use, so both apps list names identically.
func codeUnitOrder(_ lhs: String, _ rhs: String) -> Bool {
    lhs.utf16.lexicographicallyPrecedes(rhs.utf16)
}

/// Checks that each recording's numbers run 1, 2, 3, ... without a repeat or a
/// gap, and returns the first problem found.
private func findSequenceProblem(_ sorted: [Part]) -> GamePartsProblem? {
    for (offset, part) in sorted.enumerated() {
        let previous = offset > 0 ? sorted[offset - 1] : nil
        let sameRecording = previous?.label.recording == part.label.recording
        let expected = sameRecording ? (previous?.label.index ?? 0) + 1 : 1
        if sameRecording, let previous, part.label.index == previous.label.index {
            return .repeated(part.label, first: previous.name, second: part.name)
        }
        if part.label.index != expected {
            return .missing(
                PartLabel(scheme: part.label.scheme, recording: part.label.recording, index: expected)
            )
        }
    }
    return nil
}

/// Picks and orders the game parts among the file names of one folder.
public func selectGameParts(_ fileNames: [String]) -> GamePartsResult {
    var parts: [Part] = []
    var ignored: [String] = []
    for name in fileNames {
        if let part = parsePart(name) {
            parts.append(part)
        } else {
            ignored.append(name)
        }
    }
    ignored.sort(by: codeUnitOrder)

    if parts.isEmpty { return .none(ignored: ignored) }

    let schemes = Set(parts.map(\.label.scheme)).sorted()
    if schemes.count > 1 { return .invalid(.mixedSchemes(schemes)) }
    if parts.count > maxGameParts { return .invalid(.tooManyParts(count: parts.count)) }

    // The name breaks ties only so a repeated part is reported the same way
    // whatever order the directory listing came in.
    let sorted = parts.sorted { lhs, rhs in
        if lhs.label.recording != rhs.label.recording {
            return lhs.label.recording < rhs.label.recording
        }
        if lhs.label.index != rhs.label.index { return lhs.label.index < rhs.label.index }
        return codeUnitOrder(lhs.name, rhs.name)
    }
    if let problem = findSequenceProblem(sorted) { return .invalid(problem) }

    return .parts(scheme: schemes[0], parts: sorted.map(\.name), ignored: ignored)
}
